### DISCUSSION TITLE
Safe and Unsafe Carbon

### DISCUSSION THREAD

**Discussion by freedxmgxd** [carbon-language/carbon-lang#1814] [Language design]
I think that Carbon should follow the Rust to have a safe and unsafe Carbon. I know that it's impossible to be out from the Unsafe since the Interoperability with C++, but I think that a Default Safe Carbon ever is possible and verbose unsafe where it's not possible to be safe should be a design approach.

**Comment by C-BJ** [comment]
> I think that Carbon should follow the Rust to have a safe and unsafe Carbon. I know that it's impossible to be out from the Unsafe since the Interoperability with C++, but I think that a Default Safe Carbon ever is possible and verbose unsafe where it's not possible to be safe should be a design approach.

I agree with you, but security doesn't have to be the Rust way

**Comment by jonmeow** [comment]
If Rust works for you, please use it. It's a great language. If you're not sure why we're not matching Rust's approach, please read [Why not Rust?](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#why-not-rust) in the FAQ. However, [Carbon's safety strategy](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/safety_strategy.md) is different. To quote:

> Overall, Carbon is making a compromise around safety in order to give a path for C++ to evolve. C++ developers must be comfortable migrating their codebases, and able to do so in a largely automated manner. In order to achieve automated migration, Carbon cannot require fundamental redesigns of migrated C++ code. While a migration tool could in theory mark all migrated code as `unsafe`, Carbon should use a safety strategy that degrades gracefully and offers improvements for C++ code, whether migrated or not.

**Reply by freedxmgxd** [reply]
Reading the [Carbon's safety strategy](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/safety_strategy.md) helps me understand why not match Rust's approach, I really think that the unsafe blocks of Rust and the safe by default, it's a pretty elegant approach, since allow that's unsafe things are done, but demand that's the developer stay conscious about the unsafety. I will read more about safety and performance and hope I can contribute to the project.
